File: knapsack/knapsack.cpp
/*  Author:             Nikita Sandstrom
*   Last Updated:       31/10/2021
*   File Name:          knapsack.cpp
*   Kattis Attempted:   Knapsack
*/ 

#include <iostream>
#include <vector>

using namespace std;

//Function that creates the table for the Knapsack Problem
void createTable(vector<pair<int, int>>& objPairs, vector<vector<int>>& knapsackTable, int W) {
    for (int i = 0; i < objPairs.size(); i++) {
        for (int w = 0; w <= W; w++) {
            //Recurrence utilized from slides.
            if (i == 0 || w == 0) {
                knapsackTable[i][w] = 0;
            } else if (objPairs[i].second > w) {
                knapsackTable[i][w] = knapsackTable[i - 1][w];
            } else {
                knapsackTable[i][w] = max(objPairs[i].first + knapsackTable[i - 1][w - objPairs[i].second], knapsackTable[i - 1][w]);
            }
        }
    }
}

//Function that searches the table and gathers the indices for the most optimal knapsack recurrence.
void knapSackIndices(vector<vector<int>>& knapsackTable, int n, int W, vector<int>& indexResults, vector<pair<int, int>>& objPairs) {
    //base case
    if(knapsackTable[n][W] == 0) {
        return;
    } else if(knapsackTable[n-1][W] == knapsackTable[n][W]) {
        //If the value previous is the same as the one being searched at this moment, then that means this indice did not fit the weight bounds
        //or was not optimal enough
        knapSackIndices(knapsackTable, n-1, W, indexResults, objPairs);
    } else {
        //Gather the index most optimal, check back one, as well as checking up.
        //Search method is in essence a diagonal search.
        knapSackIndices(knapsackTable, n-1, W-objPairs[n].second, indexResults, objPairs);
        //sends the index to vector 
        indexResults.push_back(n-1);
    }
}

int main() {
    int valueL, valueR;
    while(cin >> valueL >> valueR) {
        //Various starting lists
        vector<pair<int, int>> totalList;
        //sets index value for results, almost a dummy value
        totalList.push_back({0,0});
        vector<vector<int>> table;
        //sets table size immediately
        table.resize(valueR + 1, vector<int>(valueL + 1));
        vector<int> indexResults;

        //gathers input for that test case
        for(int i = 0; i < valueR; i++) {
            pair<int, int> objItem;
            cin >> objItem.first >> objItem.second;
            totalList.push_back(objItem);
        }

        //Helper Functions
        createTable(totalList, table, valueL);
        knapSackIndices(table, valueR, valueL, indexResults, totalList);

        //Output
        cout << indexResults.size() << endl;
        for(int i = 0; i < indexResults.size(); i++) {
            cout << indexResults[i] << " ";
        }
        cout << endl;
    }
}

File: knapsack/knapsack.js
const readline = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout
});

var indexResults = new Array();

async function main () {
    for(var a = 0; a < 30; a++) {
        var inputInVal = await input();
        inputInVal = inputInVal.split(" ").map(Number);

        var totalList = new Array();
        totalList.push([0,0]);

        var table = new Array(inputInVal[1] + 1);

        for (var i = 0; i < inputInVal[1]; i++) {
            var objItem = await input();
            objItem = objItem.split(" ").map(Number);
            totalList.push(objItem);
        }

        for (var i = 0; i <= inputInVal[1]; i++) {
            table[i] = new Array(inputInVal[0] + 1);
            for (var w = 0; w <= inputInVal[0]; w++) {
                if (i == 0 || w == 0) {
                    table[i][w] = 0;
                } else if (totalList[i][1] > w) {
                    table[i][w] = table[i - 1][w];
                } else {
                    table[i][w] = Math.max(totalList[i][0] + table[i-1][w - totalList[i][1]], table[i - 1][w]);
                }
            }
        }
        
        knapSackIndices(table, inputInVal[1], inputInVal[0], totalList);

        console.log(indexResults.length);
        var stringBuilder = "";
        for (var i = 0; i < indexResults.length; i++) {
            stringBuilder += indexResults[i];
            if (i != indexResults.length - 1) {
                stringBuilder += " ";
            }
        }
        console.log(stringBuilder);
        indexResults = [];
    }
    
}

function knapSackIndices(knapsackTable, n, W, objPairs) {
    if (knapsackTable[n][W] == 0) {
        return;
    } else if (knapsackTable[n-1][W] == knapsackTable[n][W]) {
        knapSackIndices(knapsackTable, n-1, W, objPairs);
    } else {
        knapSackIndices(knapsackTable, n-1, W - objPairs[n][1], objPairs);
        indexResults.push(n-1);
    }
}


async function input() {    //Function to get input, bypassing NodeJS's synchronous input
    return (await readline[Symbol.asyncIterator]().next()).value;
}

main();
